src/System.Linq/src/System/Linq/Single.cs

```diff
@@ -133,27 +126,89 @@ public static TSource SingleOrDefault<TSource>(this IEnumerable<TSource> source,
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.predicate);
             }
 
-            using (IEnumerator<TSource> e = source.GetEnumerator())
+            var result = TryGetSingle(source, predicate, out bool found, out bool foundSingle);
+
+            if (!foundSingle)
             {
-                while (e.MoveNext())
+                ThrowHelper.ThrowMoreThanOneMatchException();
+
+                return default;
+            }
+
+            return result;
+        }
+
+        private static TSource TryGetSingle<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate, out bool found, out bool foundSingle)
```

Comment:
Since everyone uses `foundSingle` the same way (ThrowHelper.ThrowMoreThanOneMatchException();) you can bring the throw inside.

The method should follow the Try pattern correctly, though, and flip `out bool found` and the return value, to

```C#
private static bool TryGetSingle<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate, out TSource result)
{
    ...
}
```

Reply:
Will do. I looked at the implementation of the TryGetFirst() method which was already there and that had the same pattern. Want me to flip that one around as well?